diff --git a/modules/sd_models.py b/modules/sd_models.py
index 76940ecc..b07d84b6 100644
--- a/modules/sd_models.py
+++ b/modules/sd_models.py
@@ -482,6 +482,7 @@ def forge_model_reload():
 
     if model_data.sd_model:
         model_data.sd_model = None
+        model_data.forge_hash = None
         memory_management.unload_all_models()
         memory_management.soft_empty_cache()
         gc.collect()
