diff --git a/modules_forge/patch_basic.py b/modules_forge/patch_basic.py
index 822e2838..5893efad 100644
--- a/modules_forge/patch_basic.py
+++ b/modules_forge/patch_basic.py
@@ -39,18 +39,18 @@ def build_loaded(module, loader_name):
         except Exception as e:
             result = None
             exp = str(e) + '\n'
-            for path in list(args) + list(kwargs.values()):
-                if isinstance(path, str):
-                    if os.path.exists(path):
-                        exp += f'File corrupted: {path} \n'
-                        corrupted_backup_file = path + '.corrupted'
-                        if os.path.exists(corrupted_backup_file):
-                            os.remove(corrupted_backup_file)
-                        os.replace(path, corrupted_backup_file)
-                        if os.path.exists(path):
-                            os.remove(path)
-                        exp += f'Forge has tried to move the corrupted file to {corrupted_backup_file} \n'
-                        exp += f'You may try again now and Forge will download models again. \n'
+            # for path in list(args) + list(kwargs.values()):
+            #     if isinstance(path, str):
+            #         if os.path.exists(path):
+            #             exp += f'File corrupted: {path} \n'
+            #             corrupted_backup_file = path + '.corrupted'
+            #             if os.path.exists(corrupted_backup_file):
+            #                 os.remove(corrupted_backup_file)
+            #             os.replace(path, corrupted_backup_file)
+            #             if os.path.exists(path):
+            #                 os.remove(path)
+            #             exp += f'Forge has tried to move the corrupted file to {corrupted_backup_file} \n'
+            #             exp += f'You may try again now and Forge will download models again. \n'
             raise ValueError(exp)
         return result
 
